test/test_common/network_utility.cc

```diff
@@ -1,22 +1,78 @@
 #include "network_utility.h"
 
+#include "common/common/assert.h"
+#include "common/network/address_impl.h"
+#include "utility.h"
+
+//#include <errno.h>
+
 namespace Network {
 namespace Test {
-namespace {
-const int kFirstEphemeralPort = 40152;
-const int kLastEphemeralPort = 65535;
-const int kNumEphemeralPorts = kLastEphemeralPort - kFirstEphemeralPort + 1;
-} // namespace
 
-int getUnusedPort() {
-  static int next_port = kFirstEphemeralPort;
-  static std::mutex m;
-  std::lock_guard<std::mutex> lock(m);
-  if (next_port > kLastEphemeralPort) {
-    next_port = kFirstEphemeralPort;
+Address::InstanceConstSharedPtr checkPortAvailability(Address::InstanceConstSharedPtr addrPort,
+                                                      Address::SocketType type) {
+  // Note that there isn't any point in passing in a port of zero as we don't provide
+  // a means of finding out what port the OS assigned in that case. That could be fixed
+  // if desired.
+  RELEASE_ASSERT(addrPort != nullptr);
+  RELEASE_ASSERT(addrPort->type() == Address::Type::Ip);
+  const int fd = addrPort->socket(type);
+  RELEASE_ASSERT(fd >= 0);
+  ScopedFdCloser closer(fd);
+  // Not setting REUSEADDR, therefore if the address has been recently used we won't reuse it here.
+  // However, because we're going to use the address while checking if it is available, we'll need
+  // to set REUSEADDR on listener sockets created by tests using an address validated by this means.
+  int rc = addrPort->bind(fd);
+  if (rc == 0 && type == Address::SocketType::Stream) {
+    // Try listening on the port also, if the type is TCP.
+    rc = ::listen(fd, 1);
+  }
+  int err = errno;
```

Comment:
Nit: It's also probably not ideal to combine the error handling for bind/listen this way, it would be good to know which one failed.